packages/experimental/src/components/SidePanel/SidePanel.js

```diff
@@ -0,0 +1,585 @@
+  let cutOut = arr.splice(originalPosition, 1)[0];
+  arr.splice(newPosition, 0, cutOut);
+  return arr;
+};
+
+const SidePanel = ({
+  open,
+  setOpen,
+  placement,
+  size,
+  slideIn,
+  pageContentSelector,
+  theme,
+  includeOverlay,
+  titleText,
+  subtitleText,
+  actionToolbarButtons,
+  children,
+  condensed,
+  primaryActions,
+  animateTitle,
+  currentStep,
+  onNavigationBack,
+  selectorPrimaryFocus,
+}) => {
+  const [shouldRender, setRender] = useState(open);
+  const [animationComplete, setAnimationComplete] = useState(false);
+  const [primaryPanelActions, setPrimaryPanelActions] = useState([]);
+  const sidePanelRef = useRef();
+  const sidePanelTitleRef = useRef();
+  const sidePanelOverlayRef = useRef();
+  const startTrapRef = useRef();
+  const endTrapRef = useRef();
+  const sidePanelInnerRef = useRef();
+  const sidePanelCloseRef = useRef();
+
+  // set initial focus when side panel opens
+  useEffect(() => {
+    const initialFocus = (focusContainerElement) => {
+      const containerElement = focusContainerElement || sidePanelRef.current;
+      const primaryFocusElement = containerElement
+        ? containerElement.querySelector(selectorPrimaryFocus)
+        : null;
+
+      if (primaryFocusElement) {
+        return primaryFocusElement;
+      }
+
+      return sidePanelCloseRef && sidePanelCloseRef.current;
+    };
+
+    const focusButton = (focusContainerElement) => {
+      const target = initialFocus(focusContainerElement);
+      if (target) {
+        target.focus();
+      }
+    };
+    if (open && animationComplete) {
+      focusButton(sidePanelInnerRef.current);
+    }
+  }, [selectorPrimaryFocus, open, animationComplete]);
+
+  // Title animaton
+  useEffect(() => {
+    if (open && animateTitle && animationComplete) {
+      const sidePanelOuter = document.querySelector(
+        `#${prefix}-side-panel-outer`
+      );
+      sidePanelOuter &&
+        sidePanelOuter.addEventListener('scroll', () => {
+          const scrollTop = sidePanelRef.current.scrollTop;
+          const scrollBottom =
+            sidePanelRef.current.scrollHeight -
+            document.documentElement.clientHeight;
+          let scrollPercent = (scrollTop / scrollBottom) * 100;
+          if (scrollPercent >= 25) {
+            sidePanelOuter.classList.add(
+              `${prefix}-side-panel-with-condensed-header`
+            );
+          } else if (scrollPercent < 25) {
+            sidePanelOuter.classList.remove(
+              `${prefix}-side-panel-with-condensed-header`
+            );
+          }
+        });
+    }
+  }, [open, animateTitle, animationComplete]);
+
+  // error handling for primary action buttons
+  useEffect(() => {
+    if (primaryActions && primaryActions.length) {
+      if (
+        (size === 'small' || size === 'extraSmall' || size === 'medium') &&
+        primaryActions.length > 3
+      ) {
+        throw new Error(
```

Comment:
hmmm i'm not sure about throwing errors here. the first issue being translations. we're trying to avoid returning any strings from our components directly to the consumer because we're currently not doing any string translating on our end. secondly, even though i know we're trying to guide the user by providing them with established patterns i don't think we need to be yelling at them from the console about their "misusage". i think it would become very involved on our end to try to police every single use case when they should be referencing the design usage guidelines that we link in the docs.

we might want to open this up to discussion in slack with other people.

Reply:
Generally, I agree that we should probably leave it up to the consumer to just use the right configuration. _However_, I just thought it might be better to throw these errors in this case because there is very specific design guidelines as to the button combinations that you can have in this component.

Reply:
yeah that makes sense and i definitely considered the fact that these components are more pattern based and should adhere to the design guidelines. i guess i'm just unsure if that's something we should have to worry about or the consumer. and as for the translation stuff i think @lee-chase may have an opinion about it

Reply:
Hmm, food for thought that is for sure. 

I don't like leaving this up to our consumer. I think I would generate a console.error if more than three are supplied and truncate the array if the guidance is specific.

There is this suggestion here that creating your own propType function makes, which may make even more sense. I think from my recent typeScript study its a tuple so there may be better solutions out there https://stackoverflow.com/questions/42659728/react-proptypes-array-with-specific-length

Reply:
I think I'm leaning toward writing your own PropTypes validator

Reply:
Hey @lee-chase, I'm just getting around to this. I added a custom prop validator and removed this logic from the `useEffect`.